### PR TITLE
Feat/varify rich text

### PR DESCRIPTION
**Pull request by pairone**
[[中文版模板 / Chinese template](https://github.com/VisActor/VChart/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Release
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close https://github.com/VisActor/VChart/issues/1090

### 🔗 Related PR link
https://github.com/VisActor/VRender/pull/661

### 🐞 Bugserver case id
655ad0e39ea0da009564283b
655ad1e99ea0da0095642841
655ad2b09ea0da0095642845
655ad3359ea0da0095642846
655ad3949ea0da0095642847

### 💡 Background and solution

title, axis, label, tooltip:
https://github.com/VisActor/VChart/assets/40026218/44c0ce1d-9ff7-4ffe-b95e-e32c45922a1e

indicator:
https://github.com/VisActor/VChart/assets/40026218/7cbd16d1-517f-4327-acc4-53e30a0d93e5

markPoint:
screenshot-20231116-112748

markLine:
screenshot-20231116-112731

markArea:
screenshot-20231116-112706

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vchart/src/component/axis/base-axis.ts [hunk 1]
```diff
@@ -345,6 +345,8 @@ export abstract class AxisComponent<T extends ICommonAxisSpec & Record<string, a
             }
           : null,
         state: transformAxisLabelStateStyle(spec.label.state),
+        type: spec.label.type,
+        text: spec.label.text,
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/axis/base-axis.ts:349]: 这里不需要单独写，labelSpec 包含了的

image

File: packages/vchart/src/component/axis/base-axis.ts [hunk 2]
```diff
@@ -397,7 +399,10 @@ export abstract class AxisComponent<T extends ICommonAxisSpec & Record<string, a
           background: transformStateStyle(spec.title.background?.state)
         },
         pickable: spec.title.style?.pickable !== false,
-        childrenPickable: spec.title.style?.pickable !== false
+        childrenPickable: spec.title.style?.pickable !== false,
+        type: spec.title.type,
+        text: spec.title.text,
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/axis/base-axis.ts:404]: 这里也是

File: packages/vchart/src/component/label/interface.ts [hunk 3]
```diff
@@ -12,8 +13,16 @@ export interface ILabelSpec extends IComponentSpec {
   visible?: boolean;
   /** 是否支持交互。@default false */
   interactive?: boolean;
-  /** 格式化函数 */
-  formatMethod?: (text: string | string[], datum?: Datum, ctx?: ILabelFormatMethodContext) => string | string[];
+  // 文本类型：text, rich, html (区分于图元类型)
+  textType?: string;
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/label/interface.ts:17]: 这里为啥不和其他的统一，用 type

File: packages/vchart/src/component/title/interface/spec.ts [hunk 4]
```diff
@@ -102,7 +114,7 @@ export interface ITitleSpec extends Omit<IComponentSpec, 'orient'> {
      */
     lineClamp?: number;
     /**
-     * 富文本配置
+     * 富文本配置（暂时保留旧设置）
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/title/interface/spec.ts:117]: 加一个 @deprecated

File: packages/vchart/src/component/title/interface/spec.ts [hunk 5]
```diff
@@ -143,7 +155,7 @@ export interface ITitleSpec extends Omit<IComponentSpec, 'orient'> {
...
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/title/interface/spec.ts:158]: 将要废弃的属性记得都加上 @deprecated 标注，然后说明建议使用那个新属性代替

File: packages/vchart/src/component/tooltip/handler/utils/common.ts [hunk 6]
```diff
@@ -130,36 +132,53 @@ export const getScale = (element: HTMLElement, boundingClientRect?: DOMRect) =>
...
 
 /** 测量 tooltip 标签文本 */
-export const measureTooltipText = (text: string, style: ITooltipTextStyle): ITooltipTextInfo => {
-  text = (text ?? '').toString();
-  let textLines: string[];
-  if (style.multiLine) {
-    textLines = text.split('\n');
-    textLines = textLines.map((line, i) => (i < textLines.length - 1 ? line + '\n' : line));
+export const measureTooltipText = (text: string | TooltipRichTextAttrs, style: ITooltipTextStyle): ITooltipTextInfo => {
+  if (typeof text === 'object' && text !== null && (text?.type === 'rich' || text?.type === 'html')) {
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/tooltip/handler/utils/common.ts:140]: 应该先判断是否是富文本类型，再判断 text，另外这个分支的代码和 else 分支的代码很多重复了，这两个分支的主区别是 textConfig, 所以只需要在分支里处理 textConfig

File: packages/vchart/src/mark/text.ts [hunk 7]
```diff
@@ -16,8 +18,24 @@ export class TextMark extends BaseMark<ITextMarkSpec> implements ITextMark {
...
     return defaultStyle;
   }
+
+  protected _initProduct(group?: string | IGroupMark) {
+    const view = this.getVGrammarView();
+
+    // 声明语法元素
+    const id = this.getProductId();
+
+    if (this.getStyle('textType') === 'rich') {
```
Comment (kkxxkk2019) [inline packages/vchart/src/mark/text.ts:33]: html 不处理吗？

File: packages/vchart/src/component/label/interface.ts (hunk 3 above)
Comment (pairone) [reply packages/vchart/src/component/label/interface.ts:18]: 这里才开始叫type，和label对应的图元类型type冲突了，就改成了textType

**Review by kkxxkk2019 — APPROVED** [review]

**Review by kkxxkk2019 — APPROVED** [review]

**Pull request merged**